File: Data Structures & Algorithms/is-anagram/submission-1.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        hash_map = {}
        for letter in s:
            if letter not in hash_map:
                hash_map[letter] = 1
            else:
                hash_map[letter] += 1

        for letter in t:
            if letter in hash_map:
                hash_map[letter] -= 1
            else:
                return False
        
        for letter in hash_map:
            if hash_map[letter] != 0:
                return False
        
        return True

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) return 0;
        if (s.length() == 1) return 1;

        int l = 0, r = 0, length = 1;

        // character, most recent index
        Map<Character, Integer> map = new HashMap<>(); 
        while (r < s.length()) {
            char c = s.charAt(r);
            if (map.containsKey(c)) {
                l = Math.max(l, map.get(c) + 1);
            }

            map.put(c, r);
            int window = r - l + 1;
            length = Math.max(window, length);
            r++;
        }
        return length;
    }
}

// alsdskjfei
// abbcdefafb
// keep extending the window if a new character is found
// shift the window and delete entry if a duplicate character is found

